src/algorithms/rsa.rs

```diff
@@ -154,123 +140,149 @@ pub fn rsa_decrypt_and_check<R: CryptoRngCore + ?Sized>(
-                unblinder = ub;
-                break;
-            }
+    debug_assert_eq!(&key.n().clone().get(), n_params.modulus());
+
+    let mut r: BoxedUint = BoxedUint::one();
+    let mut ir: Option<BoxedUint> = None;
+    while ir.is_none() {
+        r = BoxedUint::random_mod(rng, key.n());
+        if r.is_zero().into() {
+            r = BoxedUint::one();
         }
-    }
 
-    let c = {
-        let mut rpowe = r.modpow(key.e(), key.n()); // N != 0
-        let mut c = c * &rpowe;
-        c %= key.n();
+        // r^-1 (mod n)
+        ir = r.inv_mod(key.n()).into();
+    }
 
+    let blinded = {
+        // r^e (mod n)
+        let mut rpowe = pow_mod_params(&r, &BoxedUint::from(key.e()), n_params.clone());
+        // c * r^e (mod n)
+        let c = mul_mod_params(c, &rpowe, n_params.clone());
         rpowe.zeroize();
 
         c
     };
 
-    (c, unblinder)
+    (blinded, ir.unwrap())
 }
 
 /// Given an m and and unblinding factor, unblind the m.
-fn unblind(key: &impl PublicKeyParts, m: &BigUint, unblinder: &BigUint) -> BigUint {
-    (m * unblinder) % key.n()
+fn unblind(m: &BoxedUint, unblinder: &BoxedUint, n_params: BoxedMontyParams) -> BoxedUint {
+    // m * r^-1 (mod n)
+    mul_mod_params(m, unblinder, n_params)
+}
+
+/// Computes `base.pow_mod(exp, n)` with precomputed `n_params`.
+fn pow_mod_params(base: &BoxedUint, exp: &BoxedUint, n_params: BoxedMontyParams) -> BoxedUint {
+    let base = reduce(&base, n_params);
+    base.pow(exp).retrieve()
+}
+
+/// Computes `lhs.mul_mod(rhs, n)` with precomputed `n_params`.
+fn mul_mod_params(lhs: &BoxedUint, rhs: &BoxedUint, n_params: BoxedMontyParams) -> BoxedUint {
+    // TODO: nicer api in crypto-bigint?
+    let lhs = BoxedMontyForm::new(lhs.clone(), n_params.clone());
+    let rhs = BoxedMontyForm::new(rhs.clone(), n_params);
+    (lhs * rhs).retrieve()
 }
 
 /// The following (deterministic) algorithm also recovers the prime factors `p` and `q` of a modulus `n`, given the
 /// public exponent `e` and private exponent `d` using the method described in
 /// [NIST 800-56B Appendix C.2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Br2.pdf).
-pub fn recover_primes(n: &BigUint, e: &BigUint, d: &BigUint) -> Result<(BigUint, BigUint)> {
+pub fn recover_primes(
+    n: &NonZero<BoxedUint>,
+    e: u64,
+    d: &BoxedUint,
+) -> Result<(BoxedUint, BoxedUint)> {
     // Check precondition
-    let two = BigUint::from_u8(2).unwrap();
-    if e <= &two.pow(16u32) || e >= &two.pow(256u32) {
+    if e <= 2u64.pow(16) || e >= 2u64.pow(256) {
         return Err(Error::InvalidArguments);
     }
 
     // 1. Let a = (de – 1) × GCD(n – 1, de – 1).
-    let one = BigUint::one();
-    let a = Zeroizing::new((d * e - &one) * (n - &one).gcd(&(d * e - &one)));
+    let one = BoxedUint::one();
+    let e = BoxedUint::from(e);
+
+    let a1 = d * &e - &one;
+    let a2 = (n.as_ref() - &one).gcd(&(d * e - &one)).unwrap();
```

Comment:
BY-GCD left side Odd but Odd<modulus>-1 == Even in formula - [crypto-bigint/590](https://github.com/RustCrypto/crypto-bigint/issues/590)